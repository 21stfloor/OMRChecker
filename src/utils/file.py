import argparse
import json
import os
from collections import defaultdict
from csv import QUOTE_NONNUMERIC
from time import localtime, strftime

import numpy as np
import pandas as pd

from src.utils.image import ImageUtils
from src.utils.logger import logger


def load_json(path, **rest):
    try:
        with open(path, "r") as f:
            loaded = json.load(f, **rest)
    except json.decoder.JSONDecodeError as error:
        logger.critical(error)
        raise Exception(f"Error when loading json file at: '{path}'")

    return loaded


class Paths:
    def __init__(self, output_dir):
        self.output_dir = output_dir
        self.save_marked_dir = output_dir.joinpath("CheckedOMRs")
        self.image_metrics_dir = output_dir.joinpath("ImageMetrics")
        self.results_dir = output_dir.joinpath("Results")
        self.manual_dir = output_dir.joinpath("Manual")
        self.errors_dir = self.manual_dir.joinpath("ErrorFiles")
        self.multi_marked_dir = self.manual_dir.joinpath("MultiMarkedFiles")


def setup_dirs_for_paths(paths):
    logger.info("Checking Directories...")

    if not os.path.exists(paths.save_marked_dir):
        os.makedirs(paths.save_marked_dir)
    # Main output directories
    for save_output_dir in [
        paths.save_marked_dir.joinpath("stack"),
        paths.save_marked_dir.joinpath("colored"),
        paths.save_marked_dir.joinpath("_MULTI_"),
        paths.save_marked_dir.joinpath("_MULTI_", "stack"),
        paths.save_marked_dir.joinpath("_MULTI_", "colored"),
    ]:
        if not os.path.exists(save_output_dir):
            os.mkdir(save_output_dir)
            logger.info(f"Created : {save_output_dir}")

    # Image buckets
    for save_output_dir in [
        paths.manual_dir,
        paths.multi_marked_dir,
        paths.errors_dir,
    ]:
        if not os.path.exists(save_output_dir):
            logger.info(f"Created : {save_output_dir}")
            os.makedirs(save_output_dir)
            os.mkdir(save_output_dir.joinpath("colored"))

    # Non-image directories
    for save_output_dir in [
        paths.results_dir,
        paths.image_metrics_dir,
    ]:
        if not os.path.exists(save_output_dir):
            logger.info(f"Created : {save_output_dir}")
            os.makedirs(save_output_dir)


def setup_outputs_for_template(paths, template):
    # TODO: consider moving this into a class instance
    ns = argparse.Namespace()
    logger.info("Checking Files...")

    # Include current output paths
    ns.paths = paths

    ns.empty_resp = [""] * len(template.output_columns)
    ns.sheetCols = [
        "file_id",
        "input_path",
        "output_path",
        "score",
    ] + template.output_columns
    ns.OUTPUT_SET = []
    ns.files_obj = {}
    TIME_NOW_HRS = strftime("%I%p", localtime())
    ns.filesMap = {
        "Results": os.path.join(paths.results_dir, f"Results_{TIME_NOW_HRS}.csv"),
        "MultiMarked": os.path.join(paths.manual_dir, "MultiMarkedFiles.csv"),
        "Errors": os.path.join(paths.manual_dir, "ErrorFiles.csv"),
    }

    for file_key, file_name in ns.filesMap.items():
        if not os.path.exists(file_name):
            logger.info(f"Created new file: '{file_name}'")
            # moved handling of files to pandas csv writer
            ns.files_obj[file_key] = file_name
            # Create Header Columns
            pd.DataFrame([ns.sheetCols], dtype=str).to_csv(
                ns.files_obj[file_key],
                mode="a",
                quoting=QUOTE_NONNUMERIC,
                header=False,
                index=False,
            )
        else:
            logger.info(f"Present : appending to '{file_name}'")
            ns.files_obj[file_key] = open(file_name, "a")

    return ns


class SaveImageOps:
    def __init__(self, tuning_config):
        self.save_img_list = defaultdict(list)
        self.tuning_config = tuning_config
        self.save_image_level = tuning_config.outputs.save_image_level

    def append_save_image(self, key, img):
        if img is None:
            return
        if self.save_image_level >= int(key):
            self.save_img_list[key].append(img.copy())

    # TODO: call this function appropriately to enable debug stacks again
    def save_image_stacks(self, key, filename, save_marked_dir):
        config = self.tuning_config
        if self.save_image_level >= int(key) and self.save_img_list[key] != []:
            display_height, display_width = config.outputs.display_image_dimensions
            name = os.path.splitext(filename)[0]
            result = np.hstack(
                tuple(
                    [
                        ImageUtils.resize_util(img, u_height=display_height)
                        for img in self.save_img_list[key]
                    ]
                )
            )
            result = ImageUtils.resize_util(
                result,
                min(
                    len(self.save_img_list[key]) * display_width // 3,
                    int(display_width * 2.5),
                ),
            )
            ImageUtils.save_img(
                f"{save_marked_dir}stack/{name}_{str(key)}_stack.jpg", result
            )

    def reset_all_save_img(self):
        for i in range(self.save_image_level):
            self.save_img_list[i + 1] = []
