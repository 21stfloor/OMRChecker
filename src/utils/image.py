import cv2
import matplotlib.pyplot as plt
import numpy as np

from src.utils.constants import CLR_BLACK, CLR_DARK_GRAY, CLR_GRAY, CLR_WHITE, TEXT_SIZE
from src.utils.logger import logger

plt.rcParams["figure.figsize"] = (10.0, 8.0)
CLAHE_HELPER = cv2.createCLAHE(clipLimit=5.0, tileGridSize=(8, 8))


class ImageUtils:
    """A Static-only Class to hold common image processing utilities & wrappers over OpenCV functions"""

    @staticmethod
    def read_image_util(file_path, tuning_config):
        if tuning_config.outputs.show_colored_outputs:
            colored_image = cv2.imread(file_path, cv2.IMREAD_COLOR)
            gray_image = cv2.cvtColor(colored_image, cv2.COLOR_BGR2GRAY)
        else:
            gray_image = cv2.imread(file_path, cv2.IMREAD_GRAYSCALE)
            colored_image = None
        return gray_image, colored_image

    @staticmethod
    def save_img(path, final_marked):
        logger.info(f"Saving Image to '{path}'")
        cv2.imwrite(path, final_marked)

    @staticmethod
    def resize_util(img, u_width, u_height=None):
        h, w = img.shape[:2]
        if u_height is None:
            u_height = int(h * u_width / w)

        if u_height == h and u_width == w:
            # No need to resize
            return img
        return cv2.resize(img, (int(u_width), int(u_height)))

    @staticmethod
    def resize_util_h(img, u_height, u_width=None):
        if u_width is None:
            h, w = img.shape[:2]
            u_width = int(w * u_height / h)
        return cv2.resize(img, (int(u_width), int(u_height)))

    @staticmethod
    def grab_contours(cnts):
        # source: imutils package

        # if the length the contours tuple returned by cv2.findContours
        # is '2' then we are using either OpenCV v2.4, v4-beta, or
        # v4-official
        if len(cnts) == 2:
            cnts = cnts[0]

        # if the length of the contours tuple is '3' then we are using
        # either OpenCV v3, v4-pre, or v4-alpha
        elif len(cnts) == 3:
            cnts = cnts[1]

        # otherwise OpenCV has changed their cv2.findContours return
        # signature yet again and I have no idea WTH is going on
        else:
            raise Exception(
                (
                    "Contours tuple must have length 2 or 3, "
                    "otherwise OpenCV changed their cv2.findContours return "
                    "signature yet again. Refer to OpenCV's documentation "
                    "in that case"
                )
            )

        # return the actual contours array
        return cnts

    @staticmethod
    def normalize(img, alpha=0, beta=255):
        return cv2.normalize(img, alpha, beta, norm_type=cv2.NORM_MINMAX)

    @staticmethod
    def auto_canny(image, sigma=0.93):
        # compute the median of the single channel pixel intensities
        v = np.median(image)

        # apply automatic Canny edge detection using the computed median
        lower = int(max(0, (1.0 - sigma) * v))
        upper = int(min(255, (1.0 + sigma) * v))
        edged = cv2.Canny(image, lower, upper)

        # return the edged image
        return edged

    @staticmethod
    def adjust_gamma(image, gamma=1.0):
        # build a lookup table mapping the pixel values [0, 255] to
        # their adjusted gamma values
        inv_gamma = 1.0 / gamma
        table = np.array(
            [((i / 255.0) ** inv_gamma) * 255 for i in np.arange(0, 256)]
        ).astype("uint8")

        # apply gamma correction using the lookup table
        return cv2.LUT(image, table)

    @staticmethod
    def get_four_destination_points(ordered_corner_points):
        (tl, tr, br, bl) = ordered_corner_points

        # compute the width of the new image, which will be the
        width_a = np.sqrt(((br[0] - bl[0]) ** 2) + ((br[1] - bl[1]) ** 2))
        width_b = np.sqrt(((tr[0] - tl[0]) ** 2) + ((tr[1] - tl[1]) ** 2))

        max_width = max(int(width_a), int(width_b))
        # max_width = max(int(np.linalg.norm(br-bl)), int(np.linalg.norm(tr-tl)))

        # compute the height of the new image, which will be the
        height_a = np.sqrt(((tr[0] - br[0]) ** 2) + ((tr[1] - br[1]) ** 2))
        height_b = np.sqrt(((tl[0] - bl[0]) ** 2) + ((tl[1] - bl[1]) ** 2))
        max_height = max(int(height_a), int(height_b))
        # max_height = max(int(np.linalg.norm(tr-br)), int(np.linalg.norm(tl-br)))

        # now that we have the dimensions of the new image, construct
        # the set of destination points to obtain a "birds eye view",
        # (i.e. top-down view) of the image, again specifying points
        # in the top-left, top-right, bottom-right, and bottom-left
        # order
        destination_points = np.array(
            [
                [0, 0],
                [max_width - 1, 0],
                [max_width - 1, max_height - 1],
                [0, max_height - 1],
            ],
            dtype="float32",
        )
        return destination_points, max_width, max_height

    @staticmethod
    def four_point_transform(image, corner_points):
        # obtain a consistent order of the points and unpack them
        # individually
        ordered_corner_points = ImageUtils.order_four_points(
            corner_points, dtype="float32"
        )

        (
            destination_points,
            max_width,
            max_height,
        ) = ImageUtils.get_four_destination_points(ordered_corner_points)
        transform_matrix = cv2.getPerspectiveTransform(
            ordered_corner_points, destination_points
        )
        warped = cv2.warpPerspective(image, transform_matrix, (max_width, max_height))

        # return the warped image
        return warped

    @staticmethod
    def order_four_points(points, dtype="int"):
        points = np.array(points)
        rect = np.zeros((4, 2), dtype=dtype)

        # the top-left point will have the smallest sum, whereas
        # the bottom-right point will have the largest sum
        s = points.sum(axis=1)
        rect[0] = points[np.argmin(s)]
        rect[2] = points[np.argmax(s)]
        diff = np.diff(points, axis=1)
        rect[1] = points[np.argmin(diff)]
        rect[3] = points[np.argmax(diff)]

        # return the ordered coordinates (tl, tr, br, bl)
        return rect

    @staticmethod
    def validate_rect(approx):
        return len(approx) == 4 and ImageUtils.check_max_cosine(approx.reshape(4, 2))

    @staticmethod
    def get_rectangle_points(x, y, w, h):
        # order same as order_four_points: (tl, tr, br, bl)
        return np.intp(
            [
                [x, y],
                [x + w, y],
                [x + w, y + h],
                [x, y + h],
            ]
        )

    @staticmethod
    def check_max_cosine(approx):
        # assumes 4 points present
        max_cosine = 0
        min_cosine = 1.5
        for i in range(2, 5):
            cosine = abs(ImageUtils.angle(approx[i % 4], approx[i - 2], approx[i - 1]))
            max_cosine = max(cosine, max_cosine)
            min_cosine = min(cosine, min_cosine)

        if max_cosine >= 0.35:
            logger.warning("Quadrilateral is not a rectangle.")
            return False
        return True

    @staticmethod
    def angle(p_1, p_2, p_0):
        dx1 = float(p_1[0] - p_0[0])
        dy1 = float(p_1[1] - p_0[1])
        dx2 = float(p_2[0] - p_0[0])
        dy2 = float(p_2[1] - p_0[1])
        return (dx1 * dx2 + dy1 * dy2) / np.sqrt(
            (dx1 * dx1 + dy1 * dy1) * (dx2 * dx2 + dy2 * dy2) + 1e-10
        )

    @staticmethod
    def get_vstack_image_grid(debug_vstack):
        padded_hstack = [
            ImageUtils.get_padded_hstack(hstack) for hstack in debug_vstack
        ]
        return ImageUtils.get_padded_vstack(padded_hstack)

    @staticmethod
    def get_padded_hstack(hstack):
        max_height = max(image.shape[0] for image in hstack)
        padded_hstack = [
            ImageUtils.pad_image_to_height(image, max_height) for image in hstack
        ]

        return np.hstack(padded_hstack)

    @staticmethod
    def get_padded_vstack(vstack):
        max_width = max(image.shape[1] for image in vstack)
        padded_vstack = [
            ImageUtils.pad_image_to_width(image, max_width) for image in vstack
        ]

        return np.vstack(padded_vstack)

    @staticmethod
    def pad_image_to_height(image, max_height, value=CLR_WHITE):
        return cv2.copyMakeBorder(
            image,
            0,
            max_height - image.shape[0],
            0,
            0,
            cv2.BORDER_CONSTANT,
            value,
        )

    @staticmethod
    def pad_image_to_width(image, max_width, value=CLR_WHITE):
        return cv2.copyMakeBorder(
            image,
            0,
            0,
            0,
            max_width - image.shape[1],
            cv2.BORDER_CONSTANT,
            value,
        )

    def pad_image_from_center(image, padding_width, padding_height=0, value=255):
        input_width, input_height = image.shape[:2]
        bounding_box = [
            padding_width,
            padding_width + input_width,
            padding_height,
            padding_height + input_height,
        ]
        white = value * np.ones(
            (padding_width * 2 + input_width, padding_height * 2 + input_height),
            np.uint8,
        )
        white[
            bounding_box[0] : bounding_box[1], bounding_box[2] : bounding_box[3]
        ] = image

        return white, bounding_box

    @staticmethod
    def draw_box(
        image,
        position,
        box_dimensions,
        color=None,
        style="BOX_HOLLOW",
        thickness_factor=1 / 12,
        border=3,
    ):
        x, y = position
        box_w, box_h = box_dimensions

        position = (
            int(x + box_w * thickness_factor),
            int(y + box_h * thickness_factor),
        )
        position_diagonal = (
            int(x + box_w - box_w * thickness_factor),
            int(y + box_h - box_h * thickness_factor),
        )
        if style == "BOX_HOLLOW":
            if color is None:
                color = CLR_GRAY
        elif style == "BOX_FILLED":
            if color is None:
                color = CLR_DARK_GRAY
            border = -1

        cv2.rectangle(
            image,
            position,
            position_diagonal,
            color,
            border,
        )
        return position,position_diagonal

    
    @staticmethod
    def draw_text(
        image,
        text_value,
        position,
        font=cv2.FONT_HERSHEY_SIMPLEX,
        text_size=TEXT_SIZE,
        color=CLR_BLACK,
        thickness=2,
    ):
        if callable(position):
            size_x, size_y = cv2.getTextSize(
                text_value,
                font,
                text_size,
                thickness,
            )[0]
            position = position(size_x, size_y)

        cv2.putText(
            image,
            text_value,
            position,
            font,
            text_size,
            color,
            thickness,
        )

    @staticmethod
    def draw_symbol(
        image,
        symbol,
        position,
        position_diagonal,
        color=CLR_BLACK
     ):
        
        center_position = lambda size_x , size_y : (
            (position[0]+position_diagonal[0]-size_x)//2,
            (position[1]+position_diagonal[1]+size_y)//2
        )
        

        ImageUtils.draw_text(
            image,
            symbol,
            center_position,
            color=color
        )
        

        

