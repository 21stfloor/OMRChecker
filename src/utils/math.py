import math

import numpy as np

from src.utils.logger import logger


class MathUtils:
    """A Static-only Class to hold common math utilities & wrappers for easy integration with OpenCV and OMRChecker"""

    # TODO: move into math utils
    @staticmethod
    def distance(point1, point2):
        return math.hypot(point1[0] - point2[0], point1[1] - point2[1])

    @staticmethod
    def get_point_on_line_by_ratio(edge_line, length_ratio):
        start, end = edge_line
        return [
            start[0] + (end[0] - start[0]) * length_ratio,
            start[1] + (end[1] - start[1]) * length_ratio,
        ]

    @staticmethod
    def order_four_points(points, dtype="int"):
        points = np.array(points)
        rect = np.zeros((4, 2), dtype=dtype)

        # the top-left point will have the smallest sum, whereas
        # the bottom-right point will have the largest sum
        s = points.sum(axis=1)
        rect[0] = points[np.argmin(s)]
        rect[2] = points[np.argmax(s)]
        diff = np.diff(points, axis=1)
        rect[1] = points[np.argmin(diff)]
        rect[3] = points[np.argmax(diff)]

        # return the ordered coordinates (tl, tr, br, bl)
        return rect

    @staticmethod
    def validate_rect(approx):
        return len(approx) == 4 and MathUtils.check_max_cosine(approx.reshape(4, 2))

    @staticmethod
    def get_rectangle_points(x, y, w, h):
        # order same as order_four_points: (tl, tr, br, bl)
        return np.intp(
            [
                [x, y],
                [x + w, y],
                [x + w, y + h],
                [x, y + h],
            ]
        )

    @staticmethod
    def check_max_cosine(approx):
        # assumes 4 points present
        max_cosine = 0
        min_cosine = 1.5
        for i in range(2, 5):
            cosine = abs(MathUtils.angle(approx[i % 4], approx[i - 2], approx[i - 1]))
            max_cosine = max(cosine, max_cosine)
            min_cosine = min(cosine, min_cosine)

        if max_cosine >= 0.35:
            logger.warning("Quadrilateral is not a rectangle.")
            return False
        return True

    @staticmethod
    def angle(p_1, p_2, p_0):
        dx1 = float(p_1[0] - p_0[0])
        dy1 = float(p_1[1] - p_0[1])
        dx2 = float(p_2[0] - p_0[0])
        dy2 = float(p_2[1] - p_0[1])
        return (dx1 * dx2 + dy1 * dy2) / np.sqrt(
            (dx1 * dx1 + dy1 * dy1) * (dx2 * dx2 + dy2 * dy2) + 1e-10
        )
