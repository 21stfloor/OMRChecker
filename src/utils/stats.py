from src.utils.parsing import default_dump


class StatsByLabel:
    def __init__(self, *labels):
        self.label_counts = {label: 0 for label in labels}

    def push(self, label, number=1):
        if label not in self.label_counts:
            raise Exception(
                f"Unknown label passed to stats by label: {label}, allowed labels: {self.label_counts.keys()}"
            )
            # self.label_counts[label] = []

        self.label_counts[label] += number

    def to_json(self):
        return {
            key: default_dump(getattr(self, key))
            for key in [
                "label_counts",
            ]
        }
