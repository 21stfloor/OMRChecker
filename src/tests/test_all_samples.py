import os
import shutil
from glob import glob

from src.tests.utils import run_entry_point, setup_mocker_patches


def read_file(path):
    with open(path) as file:
        return file.read()


def run_sample(mocker, sample_path):
    setup_mocker_patches(mocker)

    input_path = os.path.join("samples", sample_path)
    output_dir = os.path.join("outputs", sample_path)
    if os.path.exists(output_dir):
        print(
            f"Warning: output directory already exists: {output_dir}. This may affect the test execution."
        )

    run_entry_point(input_path, output_dir)

    sample_outputs = extract_sample_outputs(output_dir)

    print(f"Note: removing output directory: {output_dir}")
    shutil.rmtree(output_dir)

    return sample_outputs


EXT = "*.csv"


def extract_sample_outputs(output_dir):
    sample_outputs = {}
    for _dir, _subdir, _files in os.walk(output_dir):
        for file in glob(os.path.join(_dir, EXT)):
            relative_path = os.path.relpath(file, output_dir)
            sample_outputs[relative_path] = read_file(file)
    return sample_outputs


def test_run_omr_marker_mobile(mocker, snapshot):
    sample_outputs = run_sample(mocker, "1-mobile-camera")
    assert snapshot == sample_outputs


def test_run_omr_marker_scanned(mocker, snapshot):
    sample_outputs = run_sample(mocker, "2-omr-marker-scanned")
    assert snapshot == sample_outputs


def test_run_bonus_marking(mocker, snapshot):
    sample_outputs = run_sample(mocker, "3-answer-key/bonus-marking")
    assert snapshot == sample_outputs


def test_run_answer_key_using_csv(mocker, snapshot):
    sample_outputs = run_sample(mocker, "3-answer-key/using-csv")
    assert snapshot == sample_outputs


def test_run_answer_key_using_image(mocker, snapshot):
    sample_outputs = run_sample(mocker, "3-answer-key/using-image")
    assert snapshot == sample_outputs


def test_run_answer_key_weighted_answers(mocker, snapshot):
    sample_outputs = run_sample(mocker, "3-answer-key/weighted-answers")
    assert snapshot == sample_outputs


def test_run_crop_four_dots(mocker, snapshot):
    sample_outputs = run_sample(mocker, "experimental/1-timelines-and-dots/four-dots")
    assert snapshot == sample_outputs


def test_run_crop_two_dots_one_line(mocker, snapshot):
    sample_outputs = run_sample(mocker, "experimental/1-timelines-and-dots/four-dots")
    assert snapshot == sample_outputs


def test_run_two_lines(mocker, snapshot):
    sample_outputs = run_sample(mocker, "experimental/1-timelines-and-dots/two-lines")
    assert snapshot == sample_outputs


def test_run_template_shifts(mocker, snapshot):
    sample_outputs = run_sample(mocker, "experimental/2-template-shifts")
    assert snapshot == sample_outputs


def test_run_feature_based_alignment(mocker, snapshot):
    sample_outputs = run_sample(mocker, "experimental/3-feature-based-alignment")
    assert snapshot == sample_outputs


def test_run_community_Antibodyy(mocker, snapshot):
    sample_outputs = run_sample(mocker, "community/Antibodyy")
    assert snapshot == sample_outputs


def test_run_community_ibrahimkilic(mocker, snapshot):
    sample_outputs = run_sample(mocker, "community/ibrahimkilic")
    assert snapshot == sample_outputs


def test_run_community_Sandeep_1507(mocker, snapshot):
    sample_outputs = run_sample(mocker, "community/Sandeep-1507")
    assert snapshot == sample_outputs


def test_run_community_Shamanth(mocker, snapshot):
    sample_outputs = run_sample(mocker, "community/Shamanth")
    assert snapshot == sample_outputs


def test_run_community_UmarFarootAPS(mocker, snapshot):
    sample_outputs = run_sample(mocker, "community/UmarFarootAPS")
    assert snapshot == sample_outputs
