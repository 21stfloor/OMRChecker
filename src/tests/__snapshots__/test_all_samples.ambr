# serializer version: 1
# name: test_run_answer_key_using_csv
  dict({
    'Manual/ErrorFiles.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5"
  
    ''',
    'Manual/MultiMarkedFiles.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5"
  
    ''',
    'Results/Results_05AM.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5"
      "adrian_omr.png","samples/3-answer-key/using-csv/adrian_omr.png","outputs/3-answer-key/using-csv/CheckedOMRs/adrian_omr.png","5.0","C","E","A","B","B"
  
    ''',
  })
# ---
# name: test_run_bonus_marking
  dict({
    'Manual/ErrorFiles.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11"
  
    ''',
    'Manual/MultiMarkedFiles.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11"
  
    ''',
    'Results/Results_05AM.csv': '''
      "file_id","input_path","output_path","score","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11"
      "IMG_20201116_143512.jpg","samples/3-answer-key/bonus-marking/IMG_20201116_143512.jpg","outputs/3-answer-key/bonus-marking/CheckedOMRs/IMG_20201116_143512.jpg","33.0","B","D","C","B","D","C","BC","A","C","D","C"
      "IMG_20201116_150717658.jpg","samples/3-answer-key/bonus-marking/IMG_20201116_150717658.jpg","outputs/3-answer-key/bonus-marking/CheckedOMRs/IMG_20201116_150717658.jpg","33.0","B","D","C","B","D","C","BC","A","C","D","C"
      "IMG_20201116_150750830.jpg","samples/3-answer-key/bonus-marking/IMG_20201116_150750830.jpg","outputs/3-answer-key/bonus-marking/CheckedOMRs/IMG_20201116_150750830.jpg","6.0","A","","D","C","AC","A","D","B","C","D","D"
  
    ''',
  })
# ---
# name: test_run_omr_marker_mobile
  dict({
    'MobileCamera/Manual/ErrorFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20"
  
    ''',
    'MobileCamera/Manual/MultiMarkedFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20"
  
    ''',
    'MobileCamera/Results/Results_05AM.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20"
      "sheet1.jpg","samples/1-mobile-camera/MobileCamera/sheet1.jpg","outputs/1-mobile-camera/MobileCamera/CheckedOMRs/sheet1.jpg","0","E503110026","B","","D","B","6","11","20","7","16","B","D","C","D","A","D","B","A","C","C","D"
  
    ''',
  })
# ---
# name: test_run_omr_marker_scanned
  dict({
    'ScanBatch1/Manual/ErrorFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
  
    ''',
    'ScanBatch1/Manual/MultiMarkedFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
  
    ''',
    'ScanBatch1/Results/Results_05AM.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
      "camscanner-1.jpg","samples/2-omr-marker-scanned/ScanBatch1/camscanner-1.jpg","outputs/2-omr-marker-scanned/ScanBatch1/CheckedOMRs/camscanner-1.jpg","-4.0","E204420102","D","C","A","C","B","08","52","21","85","36","B","C","A","A","D","C","C","AD","A","A","D",""
  
    ''',
    'ScanBatch2/Manual/ErrorFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
  
    ''',
    'ScanBatch2/Manual/MultiMarkedFiles.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
  
    ''',
    'ScanBatch2/Results/Results_05AM.csv': '''
      "file_id","input_path","output_path","score","Roll","q1","q2","q3","q4","q5","q6","q7","q8","q9","q10","q11","q12","q13","q14","q15","q16","q17","q18","q19","q20","q21","q22"
      "camscanner-2.jpg","samples/2-omr-marker-scanned/ScanBatch2/camscanner-2.jpg","outputs/2-omr-marker-scanned/ScanBatch2/CheckedOMRs/camscanner-2.jpg","55.0","E204420109","C","C","B","C","C","01","19","10","10","18","D","A","D","D","D","C","C","C","C","D","B","A"
  
    ''',
  })
# ---
